### ISSUE TITLE
F# build performance

### ISSUE THREAD

**Issue by pbiggar** [darklang/dark#2952]
Jotting down some performance results from https://github.com/dotnet/fsharp/discussions/11134

[msbuild.binlog.tar.gz](https://github.com/darklang/dark/files/6013588/msbuild.binlog.tar.gz)

<details>
  <summary>Normal summary</summary>
<code>
Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Prelude -> /home/dark/app/fsharp-backend/Build/out/Prelude.dll
  LibService -> /home/dark/app/fsharp-backend/Build/out/LibService.dll
  LibExecution -> /home/dark/app/fsharp-backend/Build/out/LibExecution.dll
  Wasm -> /home/dark/app/fsharp-backend/Build/out/Wasm.dll
  Wasm (Blazor output) -> /home/dark/app/fsharp-backend/Build/out/wwwroot
  LibBackend -> /home/dark/app/fsharp-backend/Build/out/LibBackend.dll
  BwdServer -> /home/dark/app/fsharp-backend/Build/out/BwdServer.dll
  ApiServer -> /home/dark/app/fsharp-backend/Build/out/ApiServer.dll
  FuzzTests -> /home/dark/app/fsharp-backend/Build/out/FuzzTests.dll
  Tests -> /home/dark/app/fsharp-backend/Build/out/Tests.dll

Project Evaluation Performance Summary:
      246 ms  /home/dark/app/fsharp-backend/tests/FuzzTests/FuzzTests.fsproj   1 calls
      300 ms  /home/dark/app/fsharp-backend/src/Prelude/Prelude.fsproj   1 calls
      497 ms  /home/dark/app/fsharp-backend/src/LibService/LibService.fsproj   1 calls
      547 ms  /home/dark/app/fsharp-backend/src/LibBackend/LibBackend.fsproj   1 calls
      549 ms  /home/dark/app/fsharp-backend/src/LibExecution/LibExecution.fsproj   1 calls
      595 ms  /home/dark/app/fsharp-backend/tests/Tests/Tests.fsproj   1 calls
      781 ms  /home/dark/app/fsharp-backend/src/Wasm/Wasm.fsproj   1 calls
      808 ms  /home/dark/app/fsharp-backend/src/ApiServer/ApiServer.fsproj   1 calls
      813 ms  /home/dark/app/fsharp-backend/src/BwdServer/BwdServer.fsproj   1 calls

Project Performance Summary:
     4845 ms  /home/dark/app/fsharp-backend/src/Prelude/Prelude.fsproj  36 calls
                  1 ms  GetTargetFrameworks                        8 calls
                  0 ms  GetNativeManifest                          8 calls
                  0 ms  GetCopyToOutputDirectoryItems              8 calls
                  0 ms  GetCurrentProjectStaticWebAssets           3 calls
     7657 ms  /home/dark/app/fsharp-backend/src/LibService/LibService.fsproj  23 calls
                  1 ms  GetTargetFrameworks                        5 calls
                  0 ms  GetNativeManifest                          5 calls
                  0 ms  GetCopyToOutputDirectoryItems              5 calls
                  0 ms  GetCurrentProjectStaticWebAssets           2 calls
    10319 ms  /home/dark/app/fsharp-backend/src/LibExecution/LibExecution.fsproj  28 calls
                  1 ms  GetTargetFrameworks                        6 calls
                  0 ms  GetNativeManifest                          6 calls
                  0 ms  GetCurrentProjectStaticWebAssets           3 calls
                  0 ms  GetCopyToOutputDirectoryItems              6 calls
    16103 ms  /home/dark/app/fsharp-backend/src/Wasm/Wasm.fsproj   1 calls
    16984 ms  /home/dark/app/fsharp-backend/src/LibBackend/LibBackend.fsproj  19 calls
                  1 ms  GetTargetFrameworks                        4 calls
                  0 ms  GetNativeManifest                          4 calls
                  0 ms  GetCurrentProjectStaticWebAssets           2 calls
                  0 ms  GetCopyToOutputDirectoryItems              4 calls
    20889 ms  /home/dark/app/fsharp-backend/src/BwdServer/BwdServer.fsproj   9 calls
                  1 ms  GetTargetFrameworks                        2 calls
                  0 ms  GetNativeManifest                          2 calls
                  0 ms  GetCopyToOutputDirectoryItems              2 calls
    21348 ms  /home/dark/app/fsharp-backend/src/ApiServer/ApiServer.fsproj   9 calls
                  0 ms  GetTargetFrameworks                        2 calls
                  0 ms  GetNativeManifest                          2 calls
                  0 ms  GetCopyToOutputDirectoryItems              2 calls
    24731 ms  /home/dark/app/fsharp-backend/tests/FuzzTests/FuzzTests.fsproj   5 calls
                  1 ms  GetTargetFrameworks                        1 calls
                  0 ms  GetNativeManifest                          1 calls
                  0 ms  GetCopyToOutputDirectoryItems              1 calls
    29681 ms  /home/dark/app/fsharp-backend/tests/Tests/Tests.fsproj   1 calls
    30455 ms  /home/dark/app/fsharp-backend/fsdark.sln   1 calls

Target Performance Summary:
        0 ms  ResolveStaticWebAssetsInputs               3 calls
        0 ms  _PrepareForScopedCss                       3 calls
        0 ms  RazorComponentGenerate                     3 calls
        0 ms  ValidateProjects                           1 calls
        0 ms  AfterBuild                                 9 calls
        0 ms  BeforeResGen                               9 calls
        0 ms  _BlazorApplyLinkPreferencesToContent       1 calls
        0 ms  BeforeCompile                              9 calls
        0 ms  CreateCustomManifestResourceNames          9 calls
        0 ms  RazorComponentGenerateCore                 3 calls
        0 ms  CoreBuild                                  9 calls
        0 ms  AfterCompile                               9 calls
        0 ms  PrepareForRun                              9 calls
        0 ms  PrepareResourceNames                       9 calls
        0 ms  _CheckForMismatchingPlatform               1 calls
        0 ms  PrepareForRazorComponentGenerate           3 calls
        0 ms  CreateSatelliteAssemblies                  9 calls
        0 ms  GetFrameworkPaths                          9 calls
        0 ms  _CopySourceItemsToOutputDirectory          9 calls
        0 ms  _ComputeCssBundles                         3 calls
        0 ms  SetWin32ManifestProperties                 9 calls
        0 ms  _RemoveRazorDeclartionsFromCompile         3 calls
        0 ms  GetReferenceAssemblyPaths                  9 calls
        0 ms  AddSourceRevisionToInformationalVersion    9 calls
        0 ms  _ChooseAppHost                             4 calls
        0 ms  Compile                                    9 calls
        0 ms  BeforeBuild                                9 calls
        0 ms  ResGen                                     9 calls
        0 ms  GetTargetPath                              9 calls
        0 ms  BeforeResolveReferences                    9 calls
        0 ms  GenerateAssemblyInfo                       9 calls
        0 ms  AfterResGen                                9 calls
        0 ms  ComputeDependencyFileCompilerOptions       2 calls
        0 ms  PrepareProjectReferences                   9 calls
        0 ms  _AddGeneratedScopedCssFiles                3 calls
        0 ms  ValidateToolsVersions                      1 calls
        0 ms  _GenerateScopedCssFiles                    3 calls
        0 ms  _PopulateCommonStateForGetCopyToOutputDirectoryItems   9 calls
        0 ms  PrepareResources                           9 calls
        0 ms  AddImplicitDefineConstants                 9 calls
        0 ms  InitializeSourceControlInformation         9 calls
        0 ms  BuildOnlySettings                          9 calls
        0 ms  EnableIntermediateOutputPathMismatchWarning   9 calls
        0 ms  AddTransitiveFrameworkReferences           3 calls
        0 ms  _IncludeGeneratedStaticWebAssetsManifest   3 calls
        0 ms  ResolveSDKReferences                       9 calls
        1 ms  ResolveLockFileAnalyzers                   9 calls
        1 ms  _CheckForObsoleteDotNetCliToolReferences   9 calls
        1 ms  BundleScopedCssFiles                       3 calls
        1 ms  _CheckForUnsupportedHostingUsage           9 calls
        1 ms  CollectPackageReferences                   9 calls
        1 ms  _ResolveScopedCssOutputs                   3 calls
        1 ms  ExpandSDKReferences                        9 calls
        1 ms  _CheckForUnsupportedNETCoreVersion         9 calls
        1 ms  _DefaultMicrosoftNETPlatformLibrary        9 calls
        1 ms  ResolveRazorComponentInputs                3 calls
        1 ms  GetNativeManifest                          7 calls
        1 ms  _GetProjectJsonPath                        9 calls
        1 ms  GetTargetFrameworksWithPlatformForSingleTargetFramework   7 calls
        1 ms  _GetAppHostPaths                           4 calls
        1 ms  _CheckForTransitiveWindowsDesktopDependencies   3 calls
        1 ms  _CheckForLanguageAndFeatureCombinationSupport   9 calls
        1 ms  _CheckForCompileOutputs                    9 calls
        1 ms  _CheckForUnsupportedCppNETCoreVersion      9 calls
        2 ms  GetTargetPathWithTargetPlatformMoniker     9 calls
        2 ms  ResolvePackageDependenciesForBuild         9 calls
        2 ms  AfterResolveReferences                     9 calls
        2 ms  _ComputeNETCoreBuildOutputFiles            9 calls
        2 ms  IncludeTransitiveProjectReferences         9 calls
        2 ms  _GenerateSatelliteAssemblyInputs           9 calls
        2 ms  GetTargetFrameworks                        7 calls
        2 ms  _CheckForUnsupportedAppHostUsage           5 calls
        2 ms  CopyIndexHtml                              1 calls
        3 ms  _GetCopyToOutputDirectoryItemsFromThisProject   9 calls
        3 ms  ResolveReferences                          9 calls
        3 ms  _GenerateCompileInputs                     9 calls
        3 ms  GetAssemblyAttributes                      9 calls
        3 ms  ResolveRazorGenerateInputs                 3 calls
        3 ms  _GenerateRuntimeConfigurationFilesInputCache   4 calls
        3 ms  CoreResGen                                 9 calls
        3 ms  _AddScopedCssBundles                       3 calls
        3 ms  ResolveStaticWebAssetsConfiguration        3 calls
        4 ms  _ComputePackageReferencePublish            9 calls
        4 ms  _CollectAllScopedCssAssets                 3 calls
        4 ms  _PrepareForStaticWebAssets                 3 calls
        5 ms  _GetFrameworkAssemblyReferences            9 calls
        5 ms  _ComputeCssScope                           3 calls
        5 ms  ResolveScopedCssInputs                     3 calls
        5 ms  _BlockWinMDsOnUnsupportedTFMs              9 calls
        6 ms  GetAssemblyVersion                         9 calls
        7 ms  SplitResourcesByCulture                    9 calls
        7 ms  ResolveRazorConfiguration                  3 calls
        8 ms  GenerateTargetFrameworkMonikerAttribute    9 calls
        8 ms  _CreateStaticWebAssetsInputsCacheFile      3 calls
        9 ms  _CopyOutOfDateSourceItemsToOutputDirectory   5 calls
        9 ms  _ComputeUserRuntimeAssemblies              8 calls
       10 ms  ResolveLockFileCopyLocalFiles              9 calls
       10 ms  _ResolveComponentCssScopes                 3 calls
       10 ms  UpdateAspNetToFrameworkReference           9 calls
       11 ms  AssignTargetPaths                          9 calls
       15 ms  _ComputeReferenceAssemblies                8 calls
       16 ms  CreateGeneratedAssemblyInfoInputsCacheFile   9 calls
       18 ms  ResolveRuntimePackAssets                   1 calls
       18 ms  CollectFSharpDesignTimeTools               9 calls
       19 ms  _PrepareForBundling                        3 calls
       19 ms  _GetRestoreProjectStyle                    9 calls
       19 ms  GenerateStaticWebAssetsManifest            3 calls
       19 ms  ResolveLockFileReferences                  9 calls
       20 ms  ResolveReferencedProjectsStaticWebAssets   3 calls
       21 ms  ValidateSolutionConfiguration              1 calls
       21 ms  _CollectTargetFrameworkForTelemetry        9 calls
       23 ms  ApplyImplicitVersions                      9 calls
       24 ms  _SplitProjectReferencesByFileExistence     9 calls
       25 ms  _ScrambleDotnetJsFileName                  1 calls
       28 ms  IncrementalClean                           9 calls
       29 ms  PrepareForBuild                            9 calls
       35 ms  _SetEmbeddedWin32ManifestProperties        9 calls
       38 ms  _CreateAppHost                             4 calls
       45 ms  _BlazorCopyFilesToOutputDirectory          1 calls
       45 ms  _GetCopyToOutputDirectoryItemsFromTransitiveProjectReferences   9 calls
       46 ms  ResolveFrameworkReferences                 9 calls
       50 ms  CopyFilesToOutputDirectory                 9 calls
       53 ms  _CleanGetCurrentAndPriorFileWrites         9 calls
       60 ms  AssignProjectConfiguration                 9 calls
       70 ms  CheckForImplicitPackageReferenceOverrides   9 calls
       71 ms  RedirectTPReferenceToNewRedistributableLocation   9 calls
       78 ms  GetCopyToOutputDirectoryItems              9 calls
       84 ms  _CopyFilesMarkedCopyLocal                  8 calls
      115 ms  _DiscoverMvcApplicationParts               3 calls
      117 ms  _CheckForInvalidConfigurationAndPlatform   9 calls
      123 ms  GenerateFSharpTextResources                9 calls
      129 ms  CoreGenerateAssemblyInfo                   9 calls
      131 ms  CheckForDuplicateItems                     9 calls
      164 ms  RedirectFSharpCoreReferenceToNewRedistributableLocation   9 calls
      166 ms  ResolveTargetingPackAssets                 9 calls
      167 ms  SetPaketCommand                            8 calls
      179 ms  GenerateBuildRuntimeConfigurationFiles     4 calls
      184 ms  FindReferenceAssembliesForReferences       9 calls
      196 ms  ProcessFrameworkReferences                 9 calls
      238 ms  _HandlePackageFileConflicts                9 calls
      239 ms  ResolvePackageAssets                       9 calls
      366 ms  PaketRestore                               8 calls
      414 ms  _ResolveBlazorWasmOutputs                  1 calls
      469 ms  ResolveCurrentProjectStaticWebAssetsInputs   3 calls
      536 ms  GenerateBuildDependencyFile                8 calls
      645 ms  ResolveAssemblyReferences                  9 calls
      846 ms  _BlazorWasmPrepareForRun                   1 calls
     1148 ms  _ProcessBlazorWasmOutputs                  1 calls
     1868 ms  _GetProjectReferenceTargetFrameworkProperties   9 calls
    30429 ms  Build                                     10 calls
    36567 ms  CoreCompile                                9 calls
    106142 ms  ResolveProjectReferences                   9 calls

Task Performance Summary:
        1 ms  Delete                                     9 calls
        2 ms  FindInList                                 2 calls
        2 ms  AssignCulture                              9 calls
        3 ms  CreateProperty                             8 calls
        3 ms  ResolveAllScopedCssAssets                  3 calls
        3 ms  ComputeCssScope                            3 calls
        3 ms  DiscoverDefaultScopedCssItems              3 calls
        4 ms  ValidateStaticWebAssetsUniquePaths         3 calls
        4 ms  RemoveDuplicates                          18 calls
        4 ms  BlazorWriteSatelliteAssemblyFile           1 calls
        4 ms  FindAppConfigFile                          9 calls
        4 ms  GetAssemblyVersion                         9 calls
        5 ms  ConvertToAbsolutePath                      9 calls
        7 ms  AssignTargetPath                          54 calls
        7 ms  CheckIfPackageReferenceShouldBeFrameworkReference  18 calls
        8 ms  GetRestoreProjectStyleTask                 9 calls
        9 ms  ApplyCssScopes                             3 calls
       10 ms  Touch                                     11 calls
       10 ms  JoinItems                                  9 calls
       10 ms  WriteLinesToFile                          25 calls
       11 ms  Hash                                      16 calls
       13 ms  AllowEmptyTelemetry                        9 calls
       13 ms  GenerateStaticWebAssetsManifest            3 calls
       16 ms  ApplyImplicitVersions                      9 calls
       17 ms  ResolveFrameworkReferences                 9 calls
       17 ms  ResolveRuntimePackAssets                   1 calls
       19 ms  ReadLinesFromFile                         25 calls
       19 ms  GenerateBlazorWebAssemblyBootJson          1 calls
       21 ms  FindUnderPath                             45 calls
       21 ms  GetPackageDirectory                       72 calls
       25 ms  Message                                   29 calls
       26 ms  MakeDir                                   18 calls
       31 ms  GetFrameworkPath                           9 calls
       33 ms  CheckForDuplicateFrameworkReferences       9 calls
       35 ms  FSharpEmbedResourceText                    9 calls
       37 ms  CreateAppHost                              4 calls
       43 ms  FSharpEmbedResXSource                      9 calls
       44 ms  CheckForImplicitPackageReferenceOverrides   9 calls
       46 ms  ResolveAppHosts                            9 calls
       52 ms  GetFileHash                                9 calls
       53 ms  AssignProjectConfiguration                 9 calls
       64 ms  CallTarget                                18 calls
       88 ms  GetReferenceNearestTargetFrameworkTask     8 calls
       94 ms  FindAssembliesWithReferencesTo             3 calls
      100 ms  CheckForDuplicateItems                    27 calls
      107 ms  ProcessFrameworkReferences                 9 calls
      124 ms  WriteCodeFragment                          9 calls
      150 ms  Copy                                      46 calls
      162 ms  ResolveTargetingPackAssets                 9 calls
      172 ms  GenerateRuntimeConfigurationFiles          4 calls
      209 ms  ResolvePackageFileConflicts                9 calls
      226 ms  ResolvePackageAssets                       9 calls
      530 ms  GenerateDepsFile                           8 calls
      629 ms  ResolveAssemblyReference                   9 calls
      667 ms  GZipCompress                               1 calls
    36556 ms  Fsc                                        9 calls
    137948 ms  MSBuild                                   37 calls
</code></details>

<details>
  <summary>Serial summary</summary>
<code>

Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Prelude -> /home/dark/app/fsharp-backend/Build/out/Prelude.dll
  LibExecution -> /home/dark/app/fsharp-backend/Build/out/LibExecution.dll
  LibService -> /home/dark/app/fsharp-backend/Build/out/LibService.dll
  LibBackend -> /home/dark/app/fsharp-backend/Build/out/LibBackend.dll
  BwdServer -> /home/dark/app/fsharp-backend/Build/out/BwdServer.dll
  ApiServer -> /home/dark/app/fsharp-backend/Build/out/ApiServer.dll
  FuzzTests -> /home/dark/app/fsharp-backend/Build/out/FuzzTests.dll
  Tests -> /home/dark/app/fsharp-backend/Build/out/Tests.dll
  Wasm -> /home/dark/app/fsharp-backend/Build/out/Wasm.dll
  Wasm (Blazor output) -> /home/dark/app/fsharp-backend/Build/out/wwwroot

Project Evaluation Performance Summary:
      100 ms  /home/dark/app/fsharp-backend/tests/FuzzTests/FuzzTests.fsproj   1 calls
      105 ms  /home/dark/app/fsharp-backend/src/LibService/LibService.fsproj   1 calls
      108 ms  /home/dark/app/fsharp-backend/tests/Tests/Tests.fsproj   1 calls
      112 ms  /home/dark/app/fsharp-backend/src/LibExecution/LibExecution.fsproj   1 calls
      114 ms  /home/dark/app/fsharp-backend/src/LibBackend/LibBackend.fsproj   1 calls
      131 ms  /home/dark/app/fsharp-backend/src/Prelude/Prelude.fsproj   1 calls
      209 ms  /home/dark/app/fsharp-backend/src/Wasm/Wasm.fsproj   1 calls
      234 ms  /home/dark/app/fsharp-backend/src/ApiServer/ApiServer.fsproj   1 calls
      355 ms  /home/dark/app/fsharp-backend/src/BwdServer/BwdServer.fsproj   1 calls

Project Performance Summary:
     3479 ms  /home/dark/app/fsharp-backend/src/LibService/LibService.fsproj  23 calls
                  0 ms  GetTargetFrameworks                        5 calls
                  1 ms  GetNativeManifest                          5 calls
                  0 ms  GetCopyToOutputDirectoryItems              5 calls
                  1 ms  GetCurrentProjectStaticWebAssets           2 calls
     3618 ms  /home/dark/app/fsharp-backend/tests/FuzzTests/FuzzTests.fsproj   5 calls
                  0 ms  GetTargetFrameworks                        1 calls
                  1 ms  GetNativeManifest                          1 calls
                  0 ms  GetCopyToOutputDirectoryItems              1 calls
     4075 ms  /home/dark/app/fsharp-backend/src/Prelude/Prelude.fsproj  36 calls
                  1 ms  GetTargetFrameworks                        8 calls
                  1 ms  GetNativeManifest                          8 calls
                  0 ms  GetCopyToOutputDirectoryItems              8 calls
                  1 ms  GetCurrentProjectStaticWebAssets           3 calls
     4561 ms  /home/dark/app/fsharp-backend/src/ApiServer/ApiServer.fsproj   9 calls
                  9 ms  GetTargetFrameworks                        2 calls
                  1 ms  GetNativeManifest                          2 calls
                  0 ms  GetCopyToOutputDirectoryItems              2 calls
     5503 ms  /home/dark/app/fsharp-backend/src/Wasm/Wasm.fsproj   1 calls
     8366 ms  /home/dark/app/fsharp-backend/tests/Tests/Tests.fsproj   1 calls
     9483 ms  /home/dark/app/fsharp-backend/src/LibExecution/LibExecution.fsproj  28 calls
                  1 ms  GetTargetFrameworks                        6 calls
                  1 ms  GetNativeManifest                          6 calls
                  0 ms  GetCopyToOutputDirectoryItems              6 calls
                  5 ms  GetCurrentProjectStaticWebAssets           3 calls
     9553 ms  /home/dark/app/fsharp-backend/src/LibBackend/LibBackend.fsproj  19 calls
                  1 ms  GetTargetFrameworks                        4 calls
                  1 ms  GetNativeManifest                          4 calls
                  1 ms  GetCurrentProjectStaticWebAssets           2 calls
                  0 ms  GetCopyToOutputDirectoryItems              4 calls
    23553 ms  /home/dark/app/fsharp-backend/src/BwdServer/BwdServer.fsproj   9 calls
                  1 ms  GetTargetFrameworks                        2 calls
                  2 ms  GetNativeManifest                          2 calls
                  0 ms  GetCopyToOutputDirectoryItems              2 calls
    42939 ms  /home/dark/app/fsharp-backend/fsdark.sln   1 calls

Target Performance Summary:
        0 ms  ResolveStaticWebAssetsInputs               3 calls
        0 ms  RazorComponentGenerate                     3 calls
        0 ms  PrepareForRazorComponentGenerate           3 calls
        0 ms  _PrepareForScopedCss                       3 calls
        0 ms  ValidateProjects                           1 calls
        0 ms  ResGen                                     9 calls
        0 ms  PrepareResourceNames                       9 calls
        0 ms  _CheckForMismatchingPlatform               1 calls
        0 ms  BeforeCompile                              9 calls
        0 ms  _BlazorApplyLinkPreferencesToContent       1 calls
        0 ms  RazorComponentGenerateCore                 3 calls
        0 ms  BeforeResGen                               9 calls
        0 ms  _RemoveRazorDeclartionsFromCompile         3 calls
        0 ms  SetWin32ManifestProperties                 9 calls
        0 ms  _ChooseAppHost                             4 calls
        0 ms  PrepareResources                           9 calls
        0 ms  AfterResGen                                9 calls
        0 ms  ResolvePackageDependenciesForBuild         9 calls
        0 ms  CreateSatelliteAssemblies                  9 calls
        0 ms  InitializeSourceControlInformation         9 calls
        0 ms  PrepareForRun                              9 calls
        0 ms  AfterBuild                                 9 calls
        0 ms  _CopySourceItemsToOutputDirectory          9 calls
        0 ms  CreateCustomManifestResourceNames          9 calls
        0 ms  GetReferenceAssemblyPaths                  9 calls
        0 ms  AfterCompile                               9 calls
        0 ms  CoreBuild                                  9 calls
        0 ms  GetFrameworkPaths                          9 calls
        0 ms  BeforeResolveReferences                    9 calls
        0 ms  _ComputeCssBundles                         3 calls
        0 ms  AddSourceRevisionToInformationalVersion    9 calls
        0 ms  ValidateToolsVersions                      1 calls
        0 ms  _AddScopedCssBundles                       3 calls
        0 ms  PrepareProjectReferences                   9 calls
        0 ms  GenerateAssemblyInfo                       9 calls
        0 ms  _AddGeneratedScopedCssFiles                3 calls
        0 ms  BeforeBuild                                9 calls
        0 ms  GetTargetPath                              9 calls
        0 ms  AfterResolveReferences                     9 calls
        0 ms  _GenerateScopedCssFiles                    3 calls
        0 ms  Compile                                    9 calls
        0 ms  AddTransitiveFrameworkReferences           3 calls
        0 ms  ResolveRazorComponentInputs                3 calls
        0 ms  ComputeDependencyFileCompilerOptions       2 calls
        0 ms  EnableIntermediateOutputPathMismatchWarning   9 calls
        0 ms  _PopulateCommonStateForGetCopyToOutputDirectoryItems   9 calls
        0 ms  BuildOnlySettings                          9 calls
        0 ms  AddImplicitDefineConstants                 9 calls
        0 ms  _CheckForTransitiveWindowsDesktopDependencies   3 calls
        0 ms  _IncludeGeneratedStaticWebAssetsManifest   3 calls
        0 ms  ExpandSDKReferences                        9 calls
        0 ms  ResolveLockFileAnalyzers                   9 calls
        0 ms  _GetProjectJsonPath                        9 calls
        0 ms  ResolveRazorGenerateInputs                 3 calls
        0 ms  ResolveSDKReferences                       9 calls
        0 ms  _CheckForUnsupportedHostingUsage           9 calls
        0 ms  _CheckForUnsupportedNETCoreVersion         9 calls
        0 ms  BundleScopedCssFiles                       3 calls
        0 ms  CollectPackageReferences                   9 calls
        0 ms  _CheckForUnsupportedCppNETCoreVersion      9 calls
        0 ms  _CheckForObsoleteDotNetCliToolReferences   9 calls
        0 ms  _ResolveScopedCssOutputs                   3 calls
        0 ms  _DefaultMicrosoftNETPlatformLibrary        9 calls
        0 ms  _GetAppHostPaths                           4 calls
        0 ms  GetNativeManifest                          7 calls
        0 ms  GetTargetFrameworksWithPlatformForSingleTargetFramework   7 calls
        1 ms  _CheckForUnsupportedAppHostUsage           5 calls
        1 ms  _CheckForLanguageAndFeatureCombinationSupport   9 calls
        1 ms  _ComputeNETCoreBuildOutputFiles            9 calls
        1 ms  GetTargetFrameworks                        7 calls
        1 ms  _ComputePackageReferencePublish            9 calls
        1 ms  _ComputeCssScope                           3 calls
        1 ms  _CheckForCompileOutputs                    9 calls
        1 ms  GetTargetPathWithTargetPlatformMoniker     9 calls
        1 ms  _GenerateCompileInputs                     9 calls
        1 ms  ResolveScopedCssInputs                     3 calls
        1 ms  GetAssemblyAttributes                      9 calls
        1 ms  CoreResGen                                 9 calls
        2 ms  ResolveReferences                          9 calls
        2 ms  _GenerateSatelliteAssemblyInputs           9 calls
        2 ms  CopyIndexHtml                              1 calls
        2 ms  CollectFSharpDesignTimeTools               9 calls
        2 ms  GetAssemblyVersion                         9 calls
        2 ms  _GetCopyToOutputDirectoryItemsFromThisProject   9 calls
        2 ms  _CollectAllScopedCssAssets                 3 calls
        2 ms  GenerateTargetFrameworkMonikerAttribute    9 calls
        2 ms  SplitResourcesByCulture                    9 calls
        3 ms  _GenerateRuntimeConfigurationFilesInputCache   4 calls
        3 ms  _ResolveComponentCssScopes                 3 calls
        3 ms  ResolveRazorConfiguration                  3 calls
        3 ms  _GetFrameworkAssemblyReferences            9 calls
        3 ms  _PrepareForStaticWebAssets                 3 calls
        3 ms  _ComputeUserRuntimeAssemblies              8 calls
        3 ms  ResolveStaticWebAssetsConfiguration        3 calls
        3 ms  AssignTargetPaths                          9 calls
        4 ms  PrepareForBuild                            9 calls
        4 ms  _CopyOutOfDateSourceItemsToOutputDirectory   5 calls
        4 ms  ResolveLockFileCopyLocalFiles              9 calls
        5 ms  ResolveLockFileReferences                  9 calls
        5 ms  _SetEmbeddedWin32ManifestProperties        9 calls
        5 ms  UpdateAspNetToFrameworkReference           9 calls
        5 ms  ApplyImplicitVersions                      9 calls
        6 ms  CreateGeneratedAssemblyInfoInputsCacheFile   9 calls
        7 ms  _CreateStaticWebAssetsInputsCacheFile      3 calls
        7 ms  _CollectTargetFrameworkForTelemetry        9 calls
        8 ms  _GetRestoreProjectStyle                    9 calls
        8 ms  GenerateStaticWebAssetsManifest            3 calls
        8 ms  IncrementalClean                           9 calls
       10 ms  _SplitProjectReferencesByFileExistence     9 calls
       10 ms  IncludeTransitiveProjectReferences         9 calls
       11 ms  _GetCopyToOutputDirectoryItemsFromTransitiveProjectReferences   9 calls
       11 ms  _ComputeReferenceAssemblies                8 calls
       11 ms  ResolveFrameworkReferences                 9 calls
       12 ms  GenerateFSharpTextResources                9 calls
       13 ms  _BlockWinMDsOnUnsupportedTFMs              9 calls
       13 ms  _ScrambleDotnetJsFileName                  1 calls
       15 ms  _CreateAppHost                             4 calls
       16 ms  ResolveReferencedProjectsStaticWebAssets   3 calls
       17 ms  ValidateSolutionConfiguration              1 calls
       17 ms  _BlazorCopyFilesToOutputDirectory          1 calls
       19 ms  CoreGenerateAssemblyInfo                   9 calls
       19 ms  ResolveRuntimePackAssets                   1 calls
       23 ms  _PrepareForBundling                        3 calls
       25 ms  GetCopyToOutputDirectoryItems              9 calls
       26 ms  AssignProjectConfiguration                 9 calls
       32 ms  _CopyFilesMarkedCopyLocal                  8 calls
       32 ms  CheckForImplicitPackageReferenceOverrides   9 calls
       37 ms  CopyFilesToOutputDirectory                 9 calls
       40 ms  ResolveTargetingPackAssets                 9 calls
       46 ms  _CleanGetCurrentAndPriorFileWrites         9 calls
       48 ms  _CheckForInvalidConfigurationAndPlatform   9 calls
       56 ms  RedirectTPReferenceToNewRedistributableLocation   9 calls
       61 ms  _DiscoverMvcApplicationParts               3 calls
       64 ms  GenerateBuildRuntimeConfigurationFiles     4 calls
       70 ms  CheckForDuplicateItems                     9 calls
       82 ms  SetPaketCommand                            8 calls
       94 ms  _HandlePackageFileConflicts                9 calls
      104 ms  ProcessFrameworkReferences                 9 calls
      136 ms  ResolvePackageAssets                       9 calls
      151 ms  FindReferenceAssembliesForReferences       9 calls
      174 ms  RedirectFSharpCoreReferenceToNewRedistributableLocation   9 calls
      200 ms  GenerateBuildDependencyFile                8 calls
      213 ms  ResolveAssemblyReferences                  9 calls
      380 ms  _ResolveBlazorWasmOutputs                  1 calls
      411 ms  ResolveCurrentProjectStaticWebAssetsInputs   3 calls
      643 ms  _GetProjectReferenceTargetFrameworkProperties   9 calls
      800 ms  _BlazorWasmPrepareForRun                   1 calls
      964 ms  _ProcessBlazorWasmOutputs                  1 calls
     1096 ms  PaketRestore                               8 calls
    30270 ms  ResolveProjectReferences                   9 calls
    35465 ms  CoreCompile                                9 calls
    42916 ms  Build                                     10 calls

Task Performance Summary:
        0 ms  Delete                                     9 calls
        1 ms  AssignCulture                              9 calls
        1 ms  FindInList                                 2 calls
        1 ms  FindAppConfigFile                          9 calls
        1 ms  ComputeCssScope                            3 calls
        1 ms  DiscoverDefaultScopedCssItems              3 calls
        1 ms  CreateProperty                             8 calls
        1 ms  GetAssemblyVersion                         9 calls
        1 ms  ValidateStaticWebAssetsUniquePaths         3 calls
        1 ms  ResolveAllScopedCssAssets                  3 calls
        2 ms  BlazorWriteSatelliteAssemblyFile           1 calls
        2 ms  RemoveDuplicates                          18 calls
        2 ms  AssignTargetPath                          54 calls
        2 ms  ApplyCssScopes                             3 calls
        3 ms  FSharpEmbedResourceText                    9 calls
        3 ms  ConvertToAbsolutePath                      9 calls
        3 ms  ResolveFrameworkReferences                 9 calls
        3 ms  JoinItems                                  9 calls
        3 ms  GetRestoreProjectStyleTask                 9 calls
        3 ms  Touch                                     11 calls
        3 ms  ApplyImplicitVersions                      9 calls
        3 ms  FSharpEmbedResXSource                      9 calls
        4 ms  CheckIfPackageReferenceShouldBeFrameworkReference  18 calls
        4 ms  GetFrameworkPath                           9 calls
        4 ms  AllowEmptyTelemetry                        9 calls
        4 ms  Hash                                      16 calls
        5 ms  MakeDir                                   18 calls
        5 ms  GetPackageDirectory                       72 calls
        5 ms  GenerateStaticWebAssetsManifest            3 calls
        6 ms  ReadLinesFromFile                         25 calls
        7 ms  WriteLinesToFile                          25 calls
        8 ms  ResolveRuntimePackAssets                   1 calls
       11 ms  FindUnderPath                             45 calls
       12 ms  CheckForDuplicateFrameworkReferences       9 calls
       15 ms  CreateAppHost                              4 calls
       17 ms  WriteCodeFragment                          9 calls
       17 ms  CallTarget                                18 calls
       18 ms  GenerateBlazorWebAssemblyBootJson          1 calls
       19 ms  ResolveAppHosts                            9 calls
       21 ms  CheckForImplicitPackageReferenceOverrides   9 calls
       22 ms  Message                                   29 calls
       23 ms  AssignProjectConfiguration                 9 calls
       30 ms  GetReferenceNearestTargetFrameworkTask     8 calls
       33 ms  GetFileHash                                9 calls
       39 ms  ResolveTargetingPackAssets                 9 calls
       53 ms  FindAssembliesWithReferencesTo             3 calls
       59 ms  CheckForDuplicateItems                    27 calls
       61 ms  Copy                                      46 calls
       62 ms  GenerateRuntimeConfigurationFiles          4 calls
       69 ms  ProcessFrameworkReferences                 9 calls
       77 ms  ResolvePackageFileConflicts                9 calls
      130 ms  ResolvePackageAssets                       9 calls
      195 ms  GenerateDepsFile                           8 calls
      209 ms  ResolveAssemblyReference                   9 calls
      510 ms  GZipCompress                               1 calls
    35460 ms  Fsc                                        9 calls
    73761 ms  MSBuild                                   37 calls
</code></details>

**Comment by cartermp** [comment]
Playing around, and using `graphBuild` seems to work without issue here and eliminated pretty much all the MSBuild overhead on my local machine.

```
dotnet build /graphBuild:True
```

This doesn't appear to fundamentally transform the build, but the subsequent perf summaries show nearly all CPU time is now just spent in Fsc, which is good (relatively :)).

Graph build would probably be more relevant if you had a significant (e.g., 30 or more) number of projects that were inter-dependent on one another though. So if you continue to grow the codebase into more and more projects, this will keep the MSBuild overhead of discovering projects low. It's not a default option and doesn't have a `dotnet build` switch, sadly. We (Microsoft) really ought to get to the bottom of that.

Some other notes:

* Use of FSharp.Data is impacting your builds a little. `LibExecution` is pulling it in for some JSON parsing, but this does result in the F# compiler instantiating and each of the 4 type providers in that package for each project in the dependency graph. This will at at least 500ms to your compile times for each project, depending on your machine (I've observed more than that before). This impacts 5 projects. I imagine you could probably replace JSON serialization with Newtonsoft.Json for now (since System.Text.Json does weird stuff with floats as per what you've noticed).
* If you use Paket groups to separate your test dependencies from your non-test dependencies it may help with some resolution times. In general, paket is a little slower than "raw" NuGet, so if you were to move away from it that would also shave a little time, but nothing really substantial.

Beyond that, I nothing really sticks out too much to me, so we'd need to take an ETL trace of a build to see if there's anything in particular that's really impacting compile times or if it's just a case of "F# compiler needs to get faster".

**Comment by cartermp** [comment]
Also, about adding F# signature files:

Unfortunately there's no tooling to do this right now :(

You'll have to hand-author them. The benefit to this, I guess, is that you can be very choosy about what gets exposed as public, since anything you don't include in the signature file will be `internal` by default. If you do this, then starting with the compiler that will ship in about a month you'll probably notice that tooling is a little snappier than before. But it won't help build times.

**Comment by pbiggar** [comment]
> dotnet build /graphBuild:True

I tried this before a few months ago, and as I recall, only one of the projects built with this flag set. Will try it again now.

**Comment by pbiggar** [comment]
> > dotnet build /graphBuild:True
> 
> I tried this before a few months ago, and as I recall, only one of the projects built with this flag set. Will try it again now.

Well, I'm not seeing that happen anymore. However, while I do see the msbuild time drop substantially, the actual wall time doesn't change at all.

**Comment by pbiggar** [comment]
Unrelated to this but related to general performance: the building this in Docker for Mac (my default setup) is 3x slower than building this on a shiny linux box with a threadripper CPU. 5.5s vs 1.5s.

**Comment by pbiggar** [comment]
Thanks again for looking into this @cartermp. It feels like FSharp.Data will be a big win so I'll try that and report back!

**Comment by cartermp** [comment]
Re: FSharp.Data - this may not be that big of a win, but it may be noticeable. I saw between 200 and 500ms per-project on my mac, on my windows machine I noticed a 700ms difference in one run but that was the biggest.

It may also make tooling performance better not to include it either, since it's significantly less memory allocations involved if you don't have a type provider in your build graph. Memory allocations really impact tooling performance over time.

Re: mac and linux machine with a beefier CPU - not too surprising 🙂. Though I think the reality is that it shouldn't matter, there's likely a lot of things we're still doing inefficiently in the compiler that should make small changes not that big of a deal for build times.

**Comment by pbiggar** [comment]
I implemented the change to avoid FSharp.Data, and it looks like I got a 10% build-time improvement.

Numbers:
- 20s - with FSharp.Data
- 18s - without FSharp.Data

(I've changed other things so the numbers aren't comparable to the previous numbers I posted)

@cartermp I'm attaching the latest performance summary, would love some advice on other things to try!

[without-fsharp-data.txt](https://github.com/darklang/dark/files/6019501/without-fsharp-data.txt)

**Comment by pbiggar** [comment]
I'm surprised by a number of behaviours of msbuild.

For example, there seems to be no difference in the build times between saving a file, and making a small change to that file. Both lead to 18s builds when that file is in a project that 5 other projects depend on.

(For reference, the file is src/LibExecution/Interpreter.fs, and the change was changing a letter in a string)

I would typically expect small changes to be extremely cheap to build, especially when they don't change the API exposed by the project. Indeed, the same changes when using Dune to compile OCaml were extremely fast, taking <3s for a change, and 0s to save the file unchanged. This is the same project (I'm rewriting it in F#) and I'm only half way through so I'd expect the OCaml code to be about twice as many LOC.

At a glance, it seems msbuild is missing multiple features, such as hash-based rebuilds, building all files/projects  in parallel, and avoiding full rebuilds for small changes.

**Comment by cartermp** [comment]
Glad to hear removing FSharp.Data helped out!

So there are two missing improvements to the F# compiler that are affecting things the most:

1. Emitting reference assemblies
2. Lack of a compiler server

The first directly addresses the "core project changed and this rebuilt everything that depends on it" scenario. MSBuild now has (opt-in) support for simply propagating an updated build output into the output directory for projects that depend on it. It doesn't help with incremental compilation of a single project, so a small string change would rebuild the entire project. But not rebuilding everything that depends on that project would be a big deal.

The second keeps the compiler warm and caches metadata references which makes everything quite a bit faster when rebuilding successively.

As for MSBuild parallelism, it's sort of tricky, but also why the `/graphBuild` switch now exists. When you don't build with `/graphBuild`, then MSBuild determines the graph of projects as it's building them. Each time it finds a project reference, it stops building the current project and builds the dependent one. Depending on circumstances, this may launch another process since it may be parallelizable. With `/graphBuild`, it does that discovery up front and then builds in parallel as much as it can. This may not matter much for 5-10 projects, but for larger solutions it helps things.

It's also worth noting that the build manager in Visual Studio has done this for many, many years and it is a reason why building from Visual Studio has traditionally been a little faster. It also has several other heuristics in project integration tooling that prevent unnecessary work when possible. I gather you're not using VS so this doesn't really apply, but it's an interesting bit of trivia I suppose 🙂.

**Comment by njlr** [comment]
> At a glance, it seems msbuild is missing multiple features, such as hash-based rebuilds, building all files/projects in parallel, and avoiding full rebuilds for small changes.

I have also been struggling to improve the build times of an F# monorepo. They still have a few rough edges, but you might have luck with the [F# rules for Bazel](https://github.com/bazelbuild/rules_dotnet). Bazel gets you remote execution, parallelization, hash-based rebuilds, etc.

**Comment by cartermp** [comment]
Just to circle back a little bit on this, we are now working to add full support for reference assembly generation, and we'll also strive to have it turned on by default: https://github.com/dotnet/fsharp/pull/11521

It is very involved work, but necessary to add a significant degree of incrementalism to build times.

> Both lead to 18s builds when that file is in a project that 5 other projects depend on.
>
> (For reference, the file is src/LibExecution/Interpreter.fs, and the change was changing a letter in a string)

This specific behavior should get a lot better when this change is in. It will mean that the changed project is still rebuilt, but since none of the API surface area was changed, the other 5 projects would _not_ get rebuilt either. They will simply have the new `.dll` copied into their output directories. So there would still be some overhead, but since they wouldn't get recompiled, that overhead is likely to be minimal.

No estimate on when it will land just yet. It is very involved work, so we're focused on making sure it's right instead of rushing something that may not work well out into a release.

**Comment by StachuDotNet** [comment]
Folded into #5246
